rwgame/GameConfig.cpp

```diff
@@ -62,25 +60,80 @@ std::string GameConfig::getDefaultConfigPath() {
-        self->m_gameLanguage = value;
-    } else if (MATCH("input", "invert_y")) {
-        self->m_inputInvertY = atoi(value) > 0;
-    } else {
-        RW_MESSAGE("Unhandled config entry [" << section << "] " << name
-                                              << " = " << value);
-        return 0;
+std::string stripComments(const std::string &str)
+{
+    auto s = std::string(str, 0, str.find_first_of(";#"));
+    return s.erase(s.find_last_not_of(" \n\r\t")+1);
+}
+
+struct StringTranslator {
+    typedef std::string internal_type;
+    typedef std::string external_type;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        return boost::optional<external_type>(stripComments(str));
     }
+    boost::optional<internal_type> put_value(const external_type &str) {
+        return boost::optional<internal_type>(str);
+    }
+};
 
-    return 1;
+struct BoolTranslator {
+    typedef std::string internal_type;
+    typedef bool        external_type;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        return boost::optional<external_type>(std::stoi(stripComments(str)) != 0);
+    }
+    boost::optional<internal_type> put_value(const external_type &b) {
+        return boost::optional<internal_type>(b ? "1" : "0");
+    }
+};
+
+bool GameConfig::readConfig(const std::string &path) {
+    pt::ptree config, defaultConfig;
+    bool success = true;
 
-#undef MATCH
+    auto read_config = [&](const std::string &key, auto &target,
+                          const auto &defaultValue, auto &translator,
+                          bool optional=true) {
+        typedef typename std::remove_reference<decltype(target)>::type targetType;
+        defaultConfig.put(key, defaultValue, translator);
+        try {
+            target = config.get<targetType>(key, translator);
+        } catch (pt::ptree_bad_path &e) {
+            if (optional) {
+              target = defaultValue;
+            } else {
+              success = false;
+              RW_MESSAGE(e.what());
+            }
+        }
+    };
+
+    try {
+        pt::read_ini(path, config);
+    } catch (pt::ini_parser_error &e) {
+        success = false;
+        RW_MESSAGE(e.what());
+    }
+
+    auto deft = StringTranslator();
+    auto boolt = BoolTranslator();
+
+    // @todo Don't allow path seperators and relative directories
+    read_config("game.path", this->m_gamePath, "/opt/games/Grand Theft Auto 3", deft, false);
+    read_config("game.language", this->m_gameLanguage, "american", deft);
+
+    read_config("input.invert_y", this->m_inputInvertY, false, boolt);
+
+    if (!success) {
+        std::stringstream strstream;
+        pt::write_ini(strstream, defaultConfig);
+
+        RW_MESSAGE("Failed to parse configuration file (" + path + ").");
+        RW_MESSAGE("Create one looking like this at \"" + path + "\":");
+        RW_MESSAGE(strstream.str());
```

Comment:
``RW_MESSAGE`` is used for debug messages, compiled in Release these messages won't go anywhere (I've updated the wiki to mention this).

I think handling failure would make more sense in the GameBase constructor where it calls ``config.isValid()`` as there you can use ``log.error()`` to output a user-facing message before exiting.